### ISSUE TITLE
Asynchronous Functions Not Running in Parallel

### ISSUE THREAD

**Issue by kd2eom** [openfaas/faas#1272]
## My actions before raising this issue
- [x] Followed the [troubleshooting guide](https://docs.openfaas.com/deployment/troubleshooting/)
- [x] Read/searched [the docs](https://docs.openfaas.com/)
- [x] Searched [past issues](/issues)

I have a function that takes about 2 seconds to run. I deployed the function with com.openfaas.scale.max=1 and com.openfaas.scale.min=1 (guaranteeing that there will always be only 1 replica) and asynchronously called the function 10 times back-to-back with 0 delay in between calls (which means the calls should have built up in the NATS Streaming queue), noting how long it took for everything to finish. (The function returns an output via X-Callback-Url when it is finished executing).

I then did this again, only this time I set com.openfaas.scale.max=50 and com.openfaas.scale.min=50. I figured that, since I now have 50 replicas, the 10 back-to-back calls should execute concurrently in parallel, with each call executing on its own replica. Thus, the time it would take for the 10 calls to finish should be less than when I only had one replica (since in that case there was no parallelism), and I should see all 10 calls return a result at approximately the same time. However, this did not happen; in fact, the function did not seem to run in parallel at all. The results were returned to me sequentially, one at a time, and the overall time it took for the whole thing to finish processing was the same. It was as if the entire thing was still running on just one single replica.

According to the docs, increasing the number of replicas should cause the function to run in parallel for asynchronous functions; however, it does not seem to be working in my case.

I apologize if I am missing something obvious here (I am relatively new to OpenFaaS).

## Expected Behaviour

10 back-to-back calls to the function should create 10 parallel instances when there are 50 pods available.

## Current Behaviour

10 back-to-back calls to the function just run in series, seemingly on just one pod, even when there are 50 pods available. There appears to be no parallelism whatsoever.

## Steps to Reproduce (for bugs)

1. Clone https://github.com/rpi-nsl/faasbench
2. Go to faasbench/NewFunctions/facedetect/openfaas/ and run `faas-cli up --label com.openfaas.scale.max=1 --label com.openfaas.scale.min=1`
3. Go to faasbench/IsolationTests/facedetect/openfaas. Run `./results_receiver` in a terminal (this is a server which receives the return values from the asynchronous function via X-Callback-Url)
4. In a separate terminal, run `./test 10 0` while `./results_receiver` is running. This will call the function asynchronously 10 times back-to-back with a 0s delay in between. Observe the time it takes for the operation to complete.
5. Go to faasbench/NewFunctions/facedetect/openfaas/ and run `faas-cli up --label com.openfaas.scale.max=50 --label com.openfaas.scale.min=50`
6. Run `./test 10 0` again. Observe the time it takes for the operation to complete.

## Context

I am an undergraduate researcher working on a project involving OpenFaaS (some more details are in the linked GitHub repo); we need to flood the NATS Streaming queue with requests to get the function to use all available replicas at once in parallel for some tests we are running. 

## Your Environment

* FaaS-CLI version ( Full output from: `faas-cli version` ):
```
CLI:
 commit:  5fcdc692e4adf04977775ea78f79b3534a44fb36
 version: 0.8.21

Gateway
 uri:     http://127.0.0.1:31112
 version: 0.15.0
 sha:     4c12c2e2f8f3f4b4bfd740afee628d8b5524230c
 commit:  Add tests for loading the logs provider value from the env


Provider
 name:          faas-netes
 orchestration: kubernetes
 version:       0.8.1 
 sha:           ba1e9488db9a31aeec86a773537ac9d9802e6482
```

* Docker version `docker version` (e.g. Docker 17.0.05 ):
```
Client: Docker Engine - Community
 Version:           19.03.0
 API version:       1.40
 Go version:        go1.12.5
 Git commit:        aeac949
 Built:             Wed Jul 17 18:15:07 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.0
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.5
  Git commit:       aeac949
  Built:            Wed Jul 17 18:13:43 2019
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.2.6
  GitCommit:        894b81a4b802e4eb2a91d1ce216b8817763c29fb
 runc:
  Version:          1.0.0-rc8
  GitCommit:        425e105d5a03fabd737a126ad93d62a9eeede87f
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683

```

* Are you using Docker Swarm or Kubernetes (FaaS-netes)?
Kubernetes

* Operating System and version (e.g. Linux, Windows, MacOS):
Ubuntu 18.04.2

* Code example or link to GitHub repo or gist to reproduce problem:
https://github.com/rpi-nsl/faasbench/tree/master/IsolationTests/facedetect/openfaas (please see Steps to Reproduce above)

* Other diagnostic information / logs from [troubleshooting guide](https://docs.openfaas.com/deployment/troubleshooting)

## Next steps

You may [join Slack](https://docs.openfaas.com/community) for community support.

**Comment by stefanprodan** [comment]
The concurrency of asyc calls goes up with the number of queue worker replicas. You can scale up the worker with Helm https://github.com/openfaas/faas-netes/blob/master/chart/openfaas/values.yaml#L69

**Comment by kd2eom** [comment]
Oh, I see, thank you for pointing that out. So, for asynchronous function calls, it's the number of queue **_worker_** replicas that matters (since they are the ones that process the requests in the queue), not the number of replicas? Does the number of replicas not affect asynchronous functions then?

**Comment by stefanprodan** [comment]
The number of queue workers affect the concurrency level while the the number of function replicas affect the parallel processing speed. In most cases you would need to find a balance between the two. If your functions are super fast executing requests then you could just scale up the works.

**Comment by kd2eom** [comment]
Got it, thank you so much!

**Comment by welteki** [comment]
Hi @christiansicari,

Have you read this page: https://docs.openfaas.com/reference/async the section on [parallelism](https://docs.openfaas.com/reference/async/#parallelism) in particular might clear things up for you.

Both scaling the queue-worker up and increasing the `maxInflight` parameter are valid options to increase parallelism. 

If you want is to help you further please open a new issue.
Make sure to fill out the issue template so we can better understand what you are trying to build or achieve.

Regards,

Han

**Comment by christiansicari** [comment]
HI @welteki, first of all, thank you.
I read this article and this [https://www.openfaas.com/blog/limits-and-backpressure/](https://www.openfaas.com/blog/limits-and-backpressure/), and it has been hard to me join what they say.
The second article says that you can concurrently run the same function max_inflight times.
The first one says that if you have N worker and M max_inflight the concurrency is equal to M*N, but this step is quite unclear, cause if my limit is already M, I cannot exploit a N*M concurrency.
And honestly, I did not understand if the function has to manage concurrent requests or not, for example with a max_inflight=2, what happens if it is called during another execution? And then, what happens if I have N replicas of the same function?
Sorry for these many question, maybe I have to open an issue or a discussion
